### PR TITLE
EDD-16: Show errors for downloads

### PR DESCRIPTION
**Pull request by macrouch**
# Overview

### What is the feature?

Show a toast with an more info dialog where a user can cancel or retry failed downloads

# Testing

### Reproduction steps

From EDSC, provide bad download links to EDD

# Checklist

- [x] I have added automated tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have bumped the `version` field in package.json and ran `npm install`

### CODE REVIEW

File: src/app/providers/Providers/Providers.jsx
Comment (macrouch) [inline src/app/providers/Providers/Providers.jsx:1]: Missing test

File: src/app/context/__tests__/AppContext.test.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+// TODO look into why this mock isnt working
```
Comment (macrouch) [inline src/app/context/__tests__/AppContext.test.js:1]: Look into this TODO

File: src/app/main.jsx [hunk 2]
```diff
@@ -2,8 +2,11 @@ import React from 'react'
...
 import App from './App'
 
-ReactDOM.createRoot(document.getElementById('root')).render(
-  <React.StrictMode>
-    <App />
-  </React.StrictMode>
-)
+ReactDOM.createRoot(document.getElementById('root')).render(<App />)
+
+// TODO I want to leave StrictMode on, but it is resulting in multiple event listeners
```
Comment (macrouch) [inline src/app/main.jsx:7]: Look into putting strict mode back on

File: src/app/pages/Downloads/Downloads.jsx [hunk 3]
```diff
@@ -196,108 +232,178 @@ const Downloads = ({
...
+        ],
+        [
+          {
+            label: 'Restart Download',
+            isActive: true,
+            isPrimary: false,
+            callback: () => onRestartDownload(downloadId),
+            icon: FaInfoCircle
+          }
+          // {
```
Comment (macrouch) [inline src/app/pages/Downloads/Downloads.jsx:377]: Remove

File: src/app/pages/Downloads/Downloads.jsx [hunk 4]
```diff
@@ -366,6 +485,11 @@ const Downloads = ({
                     <FaCheckCircle className={styles.derivedStatusIcon} />
                   )
                 }
+                {
+                  derivedStateFromDownloads === downloadStates.error && (
```
Comment (macrouch) [inline src/app/pages/Downloads/Downloads.jsx:489]: Can we remove?

File: src/app/pages/Downloads/Downloads.jsx [hunk 5]
```diff
@@ -382,6 +506,11 @@ const Downloads = ({
                       getHumanizedDownloadStates(downloadStates.completed)
                     )
                   }
+                  {
+                    derivedStateFromDownloads === downloadStates.error && (
```
Comment (macrouch) [inline src/app/pages/Downloads/Downloads.jsx:510]: Can we remove?

File: src/app/pages/Downloads/Downloads.jsx [hunk 6]
```diff
@@ -399,7 +528,7 @@ const Downloads = ({
                     ? (
                       <>
                         {
-                          !allDownloadsPaused
+                          !allDownloadsError && (!allDownloadsPaused
```
Comment (macrouch) [inline src/app/pages/Downloads/Downloads.jsx:531]: Can we remove allDownloadsError?

File: src/app/pages/Downloads/Downloads.jsx [hunk 7]
```diff
@@ -462,4 +591,6 @@ Downloads.propTypes = {
   setHasActiveDownload: PropTypes.func.isRequired
 }
 
+Downloads.whyDidYouRender = true
```
Comment (macrouch) [inline src/app/pages/Downloads/Downloads.jsx:594]: Remove

File: src/main/eventHandlers/restartDownload.ts [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+  info,
+  webContents
+}) => {
+  const { downloadId } = info
+
+  // Set the files to pending
+  await database.updateFilesWhere({
+    downloadId
+  }, {
+    state: downloadStates.pending
```
Comment (macrouch) [inline src/main/eventHandlers/restartDownload.ts:28]: Set percent back to 0

File: src/app/context/__tests__/AppContext.test.js (hunk 1 above)
Comment (macrouch) [reply src/app/context/__tests__/AppContext.test.js:1]: Not sure how to write this test, so I left it as is

File: src/app/components/DownloadItem/DownloadItem.jsx [hunk 9]
```diff
@@ -51,11 +53,13 @@ import WaitingForLogin from '../../dialogs/InitializeDownload/WaitingForLogin'
  */
 const DownloadItem = ({
   downloadName,
+  hasErrors,
   loadingMoreFiles,
   progress,
   state,
   actionsList
 }) => {
+  console.log('🚀 ~ file: DownloadItem.jsx:62 ~ hasErrors:', hasErrors)
```
[the remark this answers is not in the document]
Comment (macrouch) [reply src/app/components/DownloadItem/DownloadItem.jsx]: :doh:

**Review by dpesall — APPROVED** [review]

File: src/app/components/ToastList/ToastList.jsx [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+import * as RadixToast from '@radix-ui/react-toast'
+
+import Toast from '../Toast/Toast'
+
+import * as styles from './ToastList.module.scss'
+
+/**
+ * @typedef {Object} ToastListProps
+ * @property {Array} toasts A list of objects to configure the toasts.
+ * @property {String} [className] An optional css class name.
```
Comment (eudoroolivares2016) [inline src/app/components/ToastList/ToastList.jsx:13]:
```suggestion
 * @property {Function} dismissToast A callback function used to dismiss the toast.
 * @property {String} [className] An optional css class name.
```

**Review by eudoroolivares2016 — APPROVED** [review]

**Pull request merged**